qutip/core/data/convert.pyx

```diff
@@ -259,63 +261,171 @@ cdef class _to:
                     _converter(convert[::-1], to_t, from_t)
         for dispatcher in self.dispatchers:
             dispatcher.rebuild_lookup()
+        for dtype in self.dtypes:
+            self._str2type[dtype.__name__.lower()] = dtype
+
+    def parse(self, dtype):
+        """
+        Return a data-layer type object given its name or the type itself.
+
+        Parameters
+        ----------
+        dtype : type, str
+            Either the (case-insensitive) name of a data-layer type or a type
+            itself.
+
+        Returns
+        -------
+        type
+            A data-layer type.
+
+        Raises
+        ------
+        TypeError
+            If ``dtype`` is neither a string nor a type.
+        ValueError
+            If ``dtype`` is a name, but no data-layer type of that name is
+            registered, or if ``dtype`` is a type, but not a known data-layer
+            type.
+        """
+        if type(dtype) is type:
+            if dtype not in self.dtypes:
+                raise ValueError(
+                    "Type is not a data-layer type: " + repr(dtype))
+            return dtype
+        elif type(dtype) is str:
+            if dtype.lower() not in self._str2type:
+                raise ValueError(
+                    "Type name is not known to the data-layer: " + repr(dtype))
+            return self._str2type[dtype.lower()]
+        raise TypeError(
+            "Invalid dtype is neither a type nor a type name: " + repr(dtype))
```

Comment:
In such inner-loop code we should be careful not to repeat method calls or `in` checks:
```python
In [1]: %load_ext Cython

In [2]: %%cython
   ...: cdef dict _KNOWN = {'dense': object, 'csr': object}
   ...: def current(dtype):
   ...:     if dtype.lower() not in _KNOWN:
   ...:         raise ValueError
   ...:     return _KNOWN[dtype.lower()]
   ...:
   ...: def alternate(dtype):
   ...:     try:
   ...:         return _KNOWN[dtype.lower()]
   ...:     except KeyError:
   ...:         raise ValueError from None
   ...:

In [3]: %timeit current('CSR')
   ...: %timeit alternate('CSR')
259 ns ± 3.57 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
159 ns ± 0.394 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: 1 - 159/259
Out[4]: 0.3861003861003861
```
So a nearly 40% speedup by not duplicating calls.